packages/langium/src/grammar/langium-grammar-validator.ts

```diff
@@ -248,33 +248,40 @@ export class LangiumGrammarValidator {
         }
         for (const rule of grammar.rules.filter(ast.isParserRule)) {
             const isDataType = isDataTypeRule(rule);
-            if (!isDataType && rule.type?.name && types.has(rule.type.name) === !!rule.infers) {
-                const keywordNode = rule.infers ? findKeywordNode(rule.$cstNode, 'infers') : findKeywordNode(rule.$cstNode, 'returns');
-                accept('error', getMessage(rule.type.name, rule.infers), {
-                    node: rule.type,
+            const isInfers = !!rule.inferredType;
+            const ruleTypeName = getTypeName(rule);
+            if (!isDataType && ruleTypeName && types.has(ruleTypeName) === isInfers) {
+                const keywordNode = isInfers ? findKeywordNode(rule.$cstNode, 'infer') : findKeywordNode(rule.$cstNode, 'returns');
+                accept('error', getMessage(ruleTypeName, isInfers), {
+                    node: rule.inferredType,
```

Comment:
This code creates a ton of false positives and errors. If the type is not explicitly defined (`types.has` returns false) and no inferred type has been declared (`isInfers = false`), the execution enters this branch, although that's a valid use case. This becomes especially bad if `rule.inferredType` is undefined, since that currently leads to errors inside of the validation collector.

Using `!!rule.inferredType` to check for inference is just in error in general. If it doesn't exist (i.e. no explicit inference has been given), one could still automatically infer a type just from the rule name. Only the existence of `returnType || dataType` makes sure that the type is not inferred.
```suggestion
            const isInfers = !rule.returnType && !rule.dataType;
            const ruleTypeName = getTypeName(rule);
            if (!isDataType && ruleTypeName && types.has(ruleTypeName) === isInfers) {
                const keywordNode = isInfers ? findKeywordNode(rule.$cstNode, 'infer') : findKeywordNode(rule.$cstNode, 'returns');
                accept('error', getMessage(ruleTypeName, isInfers), {
                    node: rule.inferredType ?? rule,
```